overlord/snapstate/storehelpers.go

```diff
@@ -519,20 +561,53 @@ func refreshCandidates(ctx context.Context, st *state.State, names []string, use
 	ignoreValidationByInstanceName := make(map[string]bool)
 	nCands := 0
 
-	addCand := func(installed *store.CurrentSnap, snapst *SnapState) {
+	enforcedSets, err := EnforcedValidationSets(st)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	addCand := func(installed *store.CurrentSnap, snapst *SnapState) error {
 		// FIXME: snaps that are not active are skipped for now
 		//        until we know what we want to do
 		if !snapst.Active {
-			return
+			return nil
 		}
 
 		if len(names) == 0 && snapst.DevMode {
 			// no auto-refresh for devmode
-			return
+			return nil
 		}
 
 		if len(names) > 0 && !strutil.SortedListContains(names, installed.InstanceName) {
-			return
+			return nil
+		}
+
+		action := &store.SnapAction{
+			Action:       "refresh",
+			SnapID:       installed.SnapID,
+			InstanceName: installed.InstanceName,
+		}
+
+		if enforcedSets != nil {
+			requiredValsets, requiredRevision, err := enforcedSets.CheckPresenceRequired(naming.Snap(installed.InstanceName))
+			// note, this errors out the entire refresh
+			if err != nil {
+				return err
+			}
+			// if the snap is already at the required revision then skip it from
+			// candidates when auto-refreshing, otherwise error out.
+			if !requiredRevision.Unset() && installed.Revision == requiredRevision {
+				if opts.IsAutoRefresh {
+					return nil
+				}
+				return fmt.Errorf("snap %q is at the required revision %s by validation sets: %s, cannot update", installed.InstanceName, installed.Revision, strings.Join(requiredValsets, ","))
```

Comment:
I think this should be a noop too? the user is not really in control of this

Reply:
Indeed, thanks.